cpp/ql/test/library-tests/dataflow/asExpr/test.cpp

```diff
@@ -0,0 +1,23 @@
+void take_const_ref_int(const int &);
+
+void test_materialize_temp_int()
+{
+  take_const_ref_int(42); // $ asExpr=42 numberOfNodes="42: 2" asIndirectExpr=42
```

Comment:
> This is saying there are two direct nodes and one indirect node for 42, right? Why are there still two direct nodes?

Yep. I'm fixing the "two direct nodes for `42`" in an upcoming PR once this one has been merged 👍

> Also we might be able to make these tags a bit clearer, though I'm not sure it's all that important for such a simple test (unless you have plans to expand it).

I don't have any plans expand it now. But I'm sure we'll find uses for this in the future. So it may make sense to improve the tags sooner rather than later - although I'd prefer to do it in a follow-up PR as I'm about to open a final PR (to fix the above issue) as soon as this one has gone in. Once that's merged I'm happy to give the tags some more love based on your input.

Reply:
Doing this later makes sense to me.

But to get my thoughts out somewhere, there are two things I think we could improve about the current tag syntax:
- treat all indirection levels consistently; as it is, we're essentially treating indirection level `0` as a special case with `asExpr`, while conflating other indirection levels into one with `asIndirectExpr`.  For the current test cases this is OK because we only use indirection levels 0 and 1.
- I'm not keen on the matched parts of tags (e.g. `asExpr` and `numberOfNodes`).  If there was more than one expr on the line, I think it could get confusing, though I appreciate `numberOfNodes` begins with the expr `toString` to mostly disambiguate that.

This is my first attempt at a better tag syntax:
```
  take_const_ref_int(42); // $ asExprInfo=expr:42,indirection:0,nodes:2 asExprInfo=expr:42,indirection:1,nodes:1
```